### PR TITLE
use reflinks for copying to platform specific dirs

### PR DESCRIPTION
**Pull request by rcerven**
* CLOUDBLD-8767

Signed-off-by: Robert Cerven <rcerven@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- n/a Python type annotations added to new code
- n/a JSON/YAML configuration changes are updated in the relevant schema
- n/a Changes to metadata also update the documentation for the metadata
- n/a Pull request has a link to an osbs-docs PR for user documentation updates
- n/a New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/tasks/binary_container_build.py [hunk 1]
```diff
@@ -275,6 +275,10 @@ def get_image_size(self, dest_tag: ImageName) -> int:
         inspect_cmd = [*self._podman_remote_cmd, 'image', 'inspect', str(dest_tag)]
         try:
             output = retries.run_cmd(inspect_cmd)
+
+            logger.info("===========output of inspect")
+            logger.info(output)
+
```
Comment (ben-alkov) [inline atomic_reactor/tasks/binary_container_build.py:281]: Is this change necessary?

Comment (rcerven) [reply atomic_reactor/tasks/binary_container_build.py:281]: no forgot to remove it

Comment (rcerven) [reply atomic_reactor/tasks/binary_container_build.py]: removed

**Review by ben-alkov — APPROVED** [review]

File: requirements.in [hunk 2]
```diff
@@ -12,3 +12,5 @@ cryptography < 3.5
 requests
 koji
 PyGObject
+reflink
+pytest-runner<=5.3.1
```
Comment (chmeliik) [inline requirements.in:16]: Shouldn't pytest-runner be in requirements-build.in? AFAIK it's a build dependency of reflink

File: atomic_reactor/dirs.py [hunk 3]
```diff
@@ -25,6 +27,10 @@
 logger = logging.getLogger(__name__)
 
 
+def reflink_copy(src, dst, *, follow_symlinks=True):
+    reflink.reflink(str(src), str(dst))
```
Comment (chmeliik) [inline atomic_reactor/dirs.py:31]: Does reflink.reflink handle permissions and metadata the same was as [shutil.copy2](https://docs.python.org/3/library/shutil.html#shutil.copy2)?

File: requirements.in (hunk 2 above)
Comment (rcerven) [reply requirements.in:16]: I've tried that script for finding build deps and it didn't find it  and according  https://gitlab.com/rubdos/pyreflink/-/blob/master/setup.py it is setup dependency ....

File: atomic_reactor/dirs.py (hunk 3 above)
Comment (rcerven) [reply atomic_reactor/dirs.py:31]: it was done as part of investigation story

reflink itself doesn't have any other parameters, it is just source and dest ....

File: requirements.in (hunk 2 above)
Comment (chmeliik) [reply requirements.in:16]: `setup_requires` = build dependency
`install_requires` = runtime dependency

Confusing, I know. And turns out setup_requires is deprecated now, but still usable https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#build-system-requirement

Not sure why the script doesn't find it :thinking: 
But pip-compile doesn't find it either, which means it is not a runtime dependency

File: atomic_reactor/dirs.py (hunk 3 above)
Comment (chmeliik) [reply atomic_reactor/dirs.py:31]: > it was done as part of investigation story

Where? I can't find it

> reflink itself doesn't have any other parameters, it is just source and dest ....

Doesn't need to be reflink params, you can do `src.resolve()`  to follow symlinks (does reflink follow symlinks?), `shutil.copymode()` to copy permissions, `shutil.copystat()` to copy metadata. Though if the copy* functions cause the reflink to be replaced with a regular file, then it would defeat the purpose

File: requirements.in (hunk 2 above)
Comment (rcerven) [reply requirements.in:16]: I think I've tried to put it into build deps but it wasn't working

File: atomic_reactor/dirs.py (hunk 3 above)
Comment (rcerven) [reply atomic_reactor/dirs.py]: added resolving also in reflink_copy method (for copytree usecase)

Comment (chmeliik) [reply atomic_reactor/dirs.py]: Shouldn't we respect the `follow_symlinks` param?

File: requirements.in (hunk 2 above)
Comment (rcerven) [reply requirements.in]: moved to build deps  and it is still working

Comment (rcerven) [reply requirements.in]: moved to build deps  and it is still working

File: setup.py [hunk 4]
```diff
@@ -37,6 +37,8 @@
         'requests',
         'koji',
         'PyGObject',
+        'reflink',
+        'pytest-runner',
```
Comment (chmeliik) [inline setup.py:41]: Now that it's in build dependencies, pytest-runner doesn't need to be here anymore

**Review by chmeliik — COMMENTED** [review]
LGTM, just one detail about `install_requires` in setup.py

**Comment by chmeliik** [comment]
waiving type-annotation, the functions is supposed to have a compatible signature to copy2 and [it's hard to annotate copy2](https://github.com/python/typeshed/blob/81be86af6e27ac0135b11c7f6697d8ae9623a034/stdlib/shutil.pyi#L58-L61)

as for coverage, we're missing `reflink_copy(src, dst, follow_symlinks=False)`

File: setup.py (hunk 4 above)
Comment (rcerven) [reply setup.py]: ok will try and test it again

**Comment by rcerven** [comment]
> waiving type-annotation, the function is supposed to have a compatible signature to copy2 and [it's hard to annotate copy2](https://github.com/python/typeshed/blob/81be86af6e27ac0135b11c7f6697d8ae9623a034/stdlib/shutil.pyi#L58-L61)
> 
> as for coverage, we're missing `reflink_copy(src, dst, follow_symlinks=False)`

well we aren't using it, it is passed by default in copytree

do you want me to have just direct test of that reflink_copy ?

**Comment by chmeliik** [comment]
> well we aren't using it, it is passed by default in copytree
> 
> do you want me to have just direct test of that reflink_copy ?

That's true, probably no need to test it separately :+1:

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

File: setup.py (hunk 4 above)
Comment (rcerven) [reply setup.py]: ok it works

**Review by mkosiarc — APPROVED** [review]

**Review by hjmodi — APPROVED** [review]

**Comment by rcerven** [comment]
rn updated: use reflinks for copying to platform specific dirs
